### PR TITLE
fix: updated the regex expression to correctly identify the table nam…

### PR DESCRIPTION
**Pull request by Naveen-Goud**
**Description**

When the user tries to add a postgres datasource and add any query, if the table name has hyphens in it, then quotes get assigned to first word after the schema and the dot. For example, public."counter"-with-db
The quotes should be applied to the whole table name. For example: public."counter-with-db"

Fixes https://github.com/appsmithorg/appsmith/issues/10631
Fixes https://github.com/appsmithorg/appsmith/issues/30692

**changes in PR:**

1.updated the regec expression to idetify the table name properly with table name consists of hypens.

2.added a test case for this scenario.

**snapshots:**
before:

Screenshot from 2024-07-30 17-15-27

After:

Screenshot from 2024-08-01 10-46-53

Hi @ajinkyakulkarni  @rohan-arthur @Nikhil-Nandagopal ,Please review this PR.

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of table names in the Postgres plugin to support a wider variety of characters, including hyphens and special characters.
	- Introduced multiple new database tables for improved testing coverage.

- **Bug Fixes**
	- Adjusted table name processing to correctly format names containing a broader range of characters.

- **Tests**
	- Added new tests to verify the structure and integrity of tables with various naming conventions.
	- Expanded existing tests to accommodate additional table structures.

### CODE REVIEW

**Comment by Naveen-Goud** [comment]
Hii    @rohan-arthur @ajinkyakulkarni  ,Could you please assign the reviewer for it.

File: app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java [hunk 1]
```diff
@@ -884,7 +884,7 @@ public Mono<DatasourceStructure> getStructure(
                                     setFragments.deleteCharAt(setFragments.length() - 1);
                                 }
 
-                                final String quotedTableName = table.getName().replaceFirst("\\.(\\w+)", ".\"$1\"");
+                                final String quotedTableName = table.getName().replaceFirst("\\.(.+)", ".\"$1\"");
```
Comment (IAmAnubhavSaini) [inline app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java:887]: 👍 

However, you should just put `[a-zA-Z0-9_\-]` in place of `\\w`. OR
split the string on first dot: `.split("\\.", 2)` and then rejoin with quotes around `[1]`.

Comment (Naveen-Goud) [reply app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java:887]: hii @IAmAnubhavSaini , I take worked on the regex based on your suggestions and tried with different regex. **"\\.(\\w+)"**  ,**"\\.([a-zA-Z0-9_\-]+)"**,
but observed that the issue is reproduciable  .

 The issue is not-reproducing only witht the above regex **"\\.(.+)",**

**Comment by Naveen-Goud** [comment]
Hii @IAmAnubhavSaini , I have updated with proper regex to identify the table names with hypens .I have also cross checked with , Its working as expected.Please review it.

thank you.

File: app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java [hunk 2]
```diff
@@ -884,7 +884,7 @@ public Mono<DatasourceStructure> getStructure(
                                     setFragments.deleteCharAt(setFragments.length() - 1);
                                 }
 
-                                final String quotedTableName = table.getName().replaceFirst("\\.(\\w+)", ".\"$1\"");
+                                final String quotedTableName = table.getName().replaceFirst("\\.([\\w\\-]+)", ".\"$1\"");
```
Comment (NilanshBansal) [inline app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java:887]: @Naveen-Goud the regex LGTM. Can you add some more test cases for different table names.
For example a test that checks for table names containing underscore and hyphens both. 
`table_name-123`

**Review by NilanshBansal — CHANGES_REQUESTED** [review]

**Comment by NilanshBansal** [comment]
@Naveen-Goud this regex does not work if the table name has dots in it.
image

**Comment by NilanshBansal** [comment]
@Naveen-Goud this regex may work out better for the table names containing dots as well.

`final String quotedTableName = table.getName().replaceFirst("\\.(.+)", ".\"$1\"");`
Any reason you chose to change this in the latest commit?

**Comment by NilanshBansal** [comment]
@Naveen-Goud the server-unit-test `PostgresPluginTest#testStructure` is failing, please check. Also, please pull the latest release and run `mvn spotless:apply` before committing the changes.
image

**Comment by Naveen-Goud** [comment]
Hii @NilanshBansal , I have updated the regex in latest commit because we want to accept the hypes(-) only , when we are using dot(.) in regex , its accepting all the special characters in table names . 

if you want me to use the prev regex ,it allows all the special chars like $,#,@,*,& in table names , If you are okay with this ,I will updated it to prev regex.

My thought was to not allow special characters other than hypens(-),underscore(_)

also we cannot have mutliple dot in table names, it leads nested structure if we use mutliple dots in table names.

**Comment by NilanshBansal** [comment]
> Hii @NilanshBansal , I have updated the regex in latest commit because we want to accept the hypes(-) only , when we are using dot(.) in regex , its accepting all the special characters in table names .
> 
> if you want me to use the prev regex ,it allows all the special chars like $,#,@,*,& in table names , If you are okay with this ,I will updated it to prev regex.
> 
> My thought was to not allow special characters other than hypens(-),underscore(_)
> 
> also we cannot have mutliple dot in table names, it leads nested structure if we use mutliple dots in table names.
@Naveen-Goud yes we should update the regex to [this](https://github.com/appsmithorg/appsmith/pull/35361#issuecomment-2302689243) to allow special characters also as supabase also supports these naming for their tables.
image

image


Please make the relevant changes along with the suggestions in https://github.com/appsmithorg/appsmith/pull/35361#discussion_r1725473592
https://github.com/appsmithorg/appsmith/pull/35361#issuecomment-2303799613
https://github.com/appsmithorg/appsmith/pull/35361#issuecomment-2302689243

Let me know once you have done the changes will re-review the PR

**Comment by Naveen-Goud** [comment]
Hii @NilanshBansal , I have updated the regex and the test case as you mentioned in the above comments.Can review now.

thank you.

**Comment by NilanshBansal** [comment]
Thanks @Naveen-Goud for the changes. I have pulled the latest changes and triggered tests on the shadow PR here: https://github.com/appsmithorg/appsmith/pull/35832
If all the tests pass, we should be good to go ahead and merge this PR in.

**Comment by NilanshBansal** [comment]
@Naveen-Goud tests are failing on the shadow PR, please check and fix.
https://github.com/appsmithorg/appsmith/pull/35832#issue-2479837508

**Comment by Naveen-Goud** [comment]
Hii @NilanshBansal , The test failing are related to cypress which are for frontend. My changes are in backend only and added Junit tests.

thank you.

**Comment by NilanshBansal** [comment]
@Naveen-Goud the backend changes can also trigger failure for certain cypress tests. Can you please confirm once if these are related to your PR or not?

**Comment by Naveen-Goud** [comment]
@NilanshBansal , those tests are not related to my PR. 
thank you.

**Comment by NilanshBansal** [comment]
@Naveen-Goud thank you for your patience, this PR has been merged! We appreciate your contribution towards making Appsmith better ❤️

**Pull request merged**